iml-agent/src/action_plugins/check_ha.rs

```diff
@@ -2,94 +2,234 @@
 // Use of this source code is governed by a MIT-style
 // license that can be found in the LICENSE file.
 
-use crate::agent_error::ImlAgentError;
+use crate::{agent_error::ImlAgentError, systemd};
 use elementtree::Element;
-use futures::{future, Future, TryFutureExt};
+use futures::try_join;
 use iml_cmd::cmd_output_success;
-use std::{collections::HashMap, fmt};
-
-/// standard:provider:ocftype (e.g. ocf:heartbeat:ZFS, or stonith:fence_ipmilan)
-#[derive(serde::Deserialize, serde::Serialize, Clone, Debug)]
-pub struct ResourceAgentType {
-    pub standard: String,         // e.g. ocf, lsb, stonith, etc..
-    pub provider: Option<String>, // e.g. heartbeat, lustre, chroma
-    pub ocftype: String,          // e.g. Lustre, ZFS
+use iml_wire_types::{
+    ComponentState, ConfigState, ResourceAgentInfo, ResourceAgentType, ServiceState,
+};
+use std::collections::HashMap;
+use tokio::fs::metadata;
+
+fn create(elem: &Element, group: Option<String>) -> ResourceAgentInfo {
+    ResourceAgentInfo {
+        agent: ResourceAgentType::new(
+            elem.get_attr("class").unwrap_or("").to_string(),
+            elem.get_attr("provider").map(|s| s.to_string()),
+            elem.get_attr("type").unwrap_or("").to_string(),
+        ),
+        group,
+        id: elem.get_attr("id").unwrap_or("").to_string(),
+        args: match elem.find("instance_attributes") {
+            None => HashMap::new(),
+            Some(e) => e
+                .find_all("nvpair")
+                .map(|nv| {
+                    (
+                        nv.get_attr("name").unwrap_or("").to_string(),
+                        nv.get_attr("value").unwrap_or("").to_string(),
+                    )
+                })
+                .collect(),
+        },
+    }
+}
+
+fn process_resource_list(output: &[u8]) -> Result<Vec<ResourceAgentInfo>, ImlAgentError> {
+    let element = Element::from_reader(output)?;
+
+    Ok(element
+        .find_all("group")
+        .flat_map(|g| {
+            let name = g.get_attr("id").unwrap_or("").to_string();
```

Comment:
Is `id` a required param? Is there a valid case for it to be missing?

Reply:
`id` is a required parameter.